src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,361 @@
+#pragma once
+
+#include <math.h>
+
+#include <algorithm>
+#include <cassert>
+#include <climits>
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <variant>
+
+#include "simeng/config/yaml/ryml.hh"
+
+namespace simeng {
+namespace config {
+
+/** An enum containing all supported data types that can be expected of a config
+ * option. */
+enum ExpectedType { Bool, Double, Float, Integer, String, UInteger, Valueless };
```

Comment:
A static cast around the enum values does the job for e.g
 ```
 size_t bool_variant_index = 0;
 static_cast<size_t>(ExpectedType::Bool) == bool_variant_index // This will evaluate to true
 
 ``

Reply:
I feel like, if you need yo do this operation, wouldn't you want to stick with the enum rather than the enum class?

Reply:
The biggest benefit of enum classes is that they don't implicitly convert to integral types and are scoped values. I do somewhat agree with the point you're making and am not opposed to using enums as I agree implicit conversion is nice to have, but you can also easily do this with implicit conversion:
```
enum color {Orange = 400}
uint8_t a = Orange;
```
 Also just so we don't run into scoping problems in the future because with normal enums you can't do:
```
enum color {Orange};
enum fruit {Orange};
```

 I'll leave this for you to decide; I'm happy with whatever you decide.

Reply:
Moved to be an enum class. I'd rather do the cast than have a less well typed implementation